pkg/replica/hash.go

```
@@ -0,0 +1,413 @@
+	Method            string `json:"method"`
+	Checksum          string `json:"checksum"`
+	ModTime           string `json:"modTime"`
+	LastHashedAt      string `json:"lastHashedAt"`
+	SilentlyCorrupted bool   `json:"silentlyCorrupted"`
+}
+
+func NewSnapshotHashJob(ctx context.Context, cancel context.CancelFunc, snapshotName string, rehash bool) *SnapshotHashJob {
+	return &SnapshotHashJob{
+		Ctx:          ctx,
+		CancelFunc:   cancel,
+		SnapshotName: snapshotName,
+		Rehash:       rehash,
+
+		SnapshotHashStatus: SnapshotHashStatus{
+			State: ProgressStateInProgress,
+		},
+	}
+}
+
+func (t *SnapshotHashJob) LockFile() (fileLock *flock.Flock, err error) {
+	defer func() {
+		if err != nil && fileLock != nil && fileLock.Path() != "" {
+			if err := os.RemoveAll(fileLock.Path()); err != nil {
+				logrus.Warnf("failed to remove lock file %v since %v", fileLock.Path(), err)
+			}
+		}
+	}()
+
+	err = os.MkdirAll(FileLockDirectory, 0755)
+	if err != nil {
+		return nil, err
+	}
+
+	fileLock = flock.New(filepath.Join(FileLockDirectory, HashLockFileName))
+
+	// Blocking lock
+	err = fileLock.Lock()
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to fetch the file lock for hashing snapshot %v", t.SnapshotName)
+	}
+
+	return fileLock, nil
+}
+
+func (t *SnapshotHashJob) UnlockFile(fileLock *flock.Flock) {
+	fileLock.Unlock()
+}
+
+func (t *SnapshotHashJob) Execute() {
+	var err error
+
+	var checksum string
+	var modTime string
+	var lastHashedAt string
+	var silentlyCorrupted bool
+
+	defer func() {
+		t.StatusLock.Lock()
+		defer t.StatusLock.Unlock()
+		t.Checksum = checksum
+		t.SilentlyCorrupted = silentlyCorrupted
+
+		if err == nil {
+			t.State = ProgressStateComplete
+
+			logrus.Infof("Snapshot %v checksum %v", t.SnapshotName, checksum)
+
+			// Checksum is not recalculated in this task, so return directly.
+			if lastHashedAt == "" {
+				return
+			}
+
+			err = SetSnapshotHashInfoToXattr(t.SnapshotName, &SnapshotXattrHashInfo{
```

Can we leave this error unhandled then directly execute `t.isModTimeRemain(modTime)`?

---

Yes, the error can be ignored. Then, next task will re-calculated and set it again.
Addressed. Thank you.